src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c

```diff
@@ -119,21 +119,23 @@ void PaPulseAudio_ReleaseOperation(PaPulseAudio_HostApiRepresentation *hostapi,
 
     while( waitOperation > 0 )
     {
-        PaPulseAudio_Lock( hostapi->mainloop );
-        pa_threaded_mainloop_wait( hostapi->mainloop );
-        PaPulseAudio_UnLock( hostapi->mainloop );
 
+        PaPulseAudio_Lock( hostapi->mainloop );
         localOperationState = pa_operation_get_state( localOperation );
 
-        // No wait if operation have been DONE or CANCELLED
-        if( localOperationState != PA_OPERATION_RUNNING)
+        if( localOperationState == PA_OPERATION_RUNNING )
+        {
+            pa_threaded_mainloop_wait( hostapi->mainloop );
```

Comment:
"This function may not be called inside the event loop thread. "

See docs for `pa_threaded_mainloop_wait` here: https://www.freedesktop.org/software/pulseaudio/doxygen/thread-mainloop_8h.html#ac96d19260567d0406350f1b3a14c4bed

Please could you confirm that this is not called from the main thread? `PaPulseAudio_UnLock`/`Lock` explicitly checks ` !pa_threaded_mainloop_in_thread( mainloop )`

Reply:
It shouldn't but I'll check. It's used on Portaudio signals and Pulseaudio callback but I should be check. Thank you for reading more carefully the docs :+1: . I don't know should it be replaces with just delay as this is taken from example.

Reply:
Ok I've added small commit that print PA_DEBUG print if we are on mainloop thread and it doesn't print so I can surely say that in test cases (this function is called something like 5 times in Portaudio life span)  it won't happen.